From 372e3b1bad47bb89848cc5e3ff62c9b921587470 Mon Sep 17 00:00:00 2001
From: Arun Ravindran <arun.ravindran@intel.com>
Date: Mon, 19 Nov 2012 16:01:12 +0200
Subject: [PORT FROM MAIN] stk: fix "set up idle mode" + "Display" text icon

BZ: 15564

According to the 3GPP 31.111 chap 6.5.4: "if the terminal
receives an icon, and either an empty or no alpha text
string is given by the UICC, than the terminal shall reject
the command with general result "Command data not understood
by terminal".

This patch fix the code to respect this point for the set up
idle mode text and Display Text proactive command.

Change-Id: Ic646f848915470f62c098498780ca8eb89204a3b
Orig-Change-Id: I58cc0b71fee7fcd76b0a4fa47e3bbc680a5c046a
Origin-change-Id: I76a1d93be47675bde87a5d9e77c9401a6d35f4a8
Origin-change-Id: I9d74334f3dbaee7d3eee885f944c46ef9be167d4
Signed-off-by: Nizar Haouati <nizar.haouati@intel.com>
Signed-off-by: Guillaume Lucas <guillaumex.lucas@intel.com>
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
Reviewed-on: http://android.intel.com:8080/76078
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Reviewed-by: Badrappan, Jeevaka <jeevaka.badrappan@intel.com>
Tested-by: Louis, NicolasX <nicolasx.louis@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 .../telephony/cat/CommandParamsFactory.java        |   27 ++++++++++++++-----
 1 files changed, 20 insertions(+), 7 deletions(-)

diff --git a/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java b/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java
index 0a21c28..95f5c5a 100644
--- a/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java
+++ b/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java
@@ -321,6 +321,14 @@ class CommandParamsFactory extends Handler {
         textMsg.isHighPriority = (cmdDet.commandQualifier & 0x01) != 0;
         textMsg.userClear = (cmdDet.commandQualifier & 0x80) != 0;
 
+        // According to 3GPP 31.111 chap 6.5.4 (ETSI TS 102 223 clause 6.5.4):
+        // If the terminal receives an icon and either an empty or no alpha
+        // text string is given by UICC, than the terminal shall reject the
+        // command with general result "Command not understood by terminal".
+        if ((iconId != null) && (textMsg.text == null)) {
+            throw new ResultException(ResultCode.CMD_DATA_NOT_UNDERSTOOD);
+        }
+
         mCmdParams = new DisplayTextParams(cmdDet, textMsg);
 
         if (iconId != null) {
@@ -355,15 +363,20 @@ class CommandParamsFactory extends Handler {
         if (ctlv != null) {
             textMsg.text = ValueParser.retrieveTextString(ctlv);
         }
-        // load icons only when text exist.
-        if (textMsg.text != null) {
-            ctlv = searchForTag(ComprehensionTlvTag.ICON_ID, ctlvs);
-            if (ctlv != null) {
-                iconId = ValueParser.retrieveIconId(ctlv);
-                textMsg.iconSelfExplanatory = iconId.selfExplanatory;
-            }
+
+        ctlv = searchForTag(ComprehensionTlvTag.ICON_ID, ctlvs);
+        if (ctlv != null) {
+            iconId = ValueParser.retrieveIconId(ctlv);
+            textMsg.iconSelfExplanatory = iconId.selfExplanatory;
         }
 
+        // According to 3GPP 31.111 chap 6.5.4 (ETSI TS 102 223 clause 6.5.4):
+        // If the terminal receives an icon and either an empty or no alpha
+        // text string is given by UICC, than the terminal shall reject the
+        // command with general result "Command not understood by terminal".
+        if ((iconId != null) && (textMsg.text == null)) {
+            throw new ResultException(ResultCode.CMD_DATA_NOT_UNDERSTOOD);
+        }
         mCmdParams = new DisplayTextParams(cmdDet, textMsg);
 
         if (iconId != null) {
-- 
1.7.4.1

